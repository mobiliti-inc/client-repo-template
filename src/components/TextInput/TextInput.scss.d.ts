// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'checkmark-label': string;
  'dark': string;
  'error': string;
  'error-message': string;
  'fades-in-out': string;
  'input-base': string;
  'input-date': string;
  'input-email': string;
  'input-file': string;
  'input-group': string;
  'input-image': string;
  'input-label': string;
  'input-number': string;
  'input-password': string;
  'input-tel': string;
  'input-text': string;
  'input-url': string;
  'password-visibility-indicator': string;
}
declare var cssExports: CssExports;
export = cssExports;
