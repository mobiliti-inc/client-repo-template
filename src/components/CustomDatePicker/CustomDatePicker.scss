@import "~variables";

.container {
  font-family: $body-font;
  display: flex;
  flex-direction: column;
}

.container input {
  height: 50px;
  width: 100%;
  font-size: 1.25rem;
  color: $black;
  padding: 0;
  border: none;
  transition: all .5s;
  border-bottom: 1px solid rgba(29, 49, 85, 0.2);
  text-transform: capitalize;
  font-family: $body-font;

  &:focus {
    outline: none;
    border-bottom: 3px solid $blue;
  }

  &::-webkit-input-placeholder {
    color: $darker-blue;
    opacity: 0.6;
  }
}

.error-border input {
  border-bottom: 3px solid $fast-car-red;
}

.error {
  color: $fast-car-red;
  height: 14px;
  font-family: $hanken-sans;
  margin-top: 5px;
  font-size: 12px;
  line-height: 14px;
}

.icon {
  position: absolute;
  background-color: $white;
  top: 20px;
  width: 35px;
  border: 1px solid #d0d6de;
  border-radius: 5px;
  padding: 5px;
}

.icon:hover {
  background-color: rgba(204, 209, 218, .4);
}

.icon:active {
  background-color: $white;
}

.prev {
  left: 5px;
}

.next {
  right: 5px;
  -webkit-transform: rotate(180deg);
  /* Safari and Chrome */
  -moz-transform: rotate(180deg);
  /* Firefox */
  -ms-transform: rotate(180deg);
  /* IE 9 */
  -o-transform: rotate(180deg);
  /* Opera */
  transform: rotate(180deg);
}

.select {
  border: none;
  background: #2a98b1;
  color: #fff;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0 20px 0 5px;
  height: 40px;
  line-height: 30px;
  margin: 0 5px 0 0;
  border-radius: 2px;
  background-image: url(../../assets/icons/chevron-down-triangle.svg);
  background-position: right center;
  background-repeat: no-repeat;
}

.select:focus {
  outline: 0;
}

.calendar {
  display: flex;
  justify-content: center;
  margin-top: -9px;
}

.input-label {
  display: block;
  font-size: 12px;
  transform: translateY(-60px);
  color: $darker-blue;
  opacity: 0.6;
  font-family: $hanken-sans;
  line-height: 14px;
  text-transform: capitalize;
  transition: all .5s;

  &--focused {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-40px);
  }
}
