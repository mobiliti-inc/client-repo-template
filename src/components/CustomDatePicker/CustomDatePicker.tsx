/* eslint no-underscore-dangle: 0 */
import React from 'react';
import moment from 'moment';
import classNames from 'classnames';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import styles from './CustomDatePicker.scss';
import { directionArrow } from '../../assets/icons';

type CustomDatePickerProps = {
	placeholder?: string,
	isValid?: boolean,
	errorMessage?: string,
	onDatePicked: (...args: any[]) => any,
	openToDate?: string,
	license?: boolean,
	selectedDate?: string,
	label?: string,
	id?: string,
	numberOfYears?: number,
	isDayBlocked?: (...args: any[]) => any,
	isOutsideRange?: (...args: any[]) => any,
	negative?: boolean
};

type CustomDatePickerState = {
	date: null,
	valid: boolean,
	error: string,
	focused: boolean
};

class CustomDatePicker extends React.PureComponent<CustomDatePickerProps, CustomDatePickerState> {
	state = {
		date: null,
		valid: true,
		error: '',
		focused: false
	};

	componentDidMount() {
		const { selectedDate } = this.props;
		if (selectedDate && selectedDate._isValid) this.setDate(selectedDate);
	}

	componentDidUpdate = (prevProps, prevState) => {
		const { openToDate, selectedDate } = this.props;
		const { focused, date } = this.state;
		if (focused && prevState.focused !== focused && openToDate && this.formatDate(openToDate)._isValid && !date) {
			this.setDate(openToDate);
		}
		if (selectedDate && prevProps.selectedDate !== selectedDate && this.formatDate(selectedDate)._isValid) {
			this.setDate(selectedDate);
		}
	}

	onDateChange = date => {
		const { onDatePicked } = this.props;
		if (date && date._isValid) {
			onDatePicked({ value: moment(date).format('MM/DD/YYYY'), error: '', valid: true });
			this.setState({ date, valid: true, error: '' });
		} else {
			onDatePicked();
			this.setState({ valid: false, error: 'Please enter a valid date' });
		}
	}

	setDate = date => this.setState({ date: this.formatDate(date), valid: true, error: '' });

	// Combine classes
	getClasses = () => classNames('container', { 'error-border': !this.state.valid || !this.props.isValid });

	formatDate = date => moment(date, 'MM/DD/YYYY');

	renderMonthElement = ({ month, onMonthSelect, onYearSelect }) => (
		<div className={styles.calendar}>
			<div>
				<select value={month.month()} onChange={e => onMonthSelect(month, e.target.value)} className={styles.select}>
					{moment.months().map((label, value) => (
						<option value={value} key={label}>
							{label}
						</option>
					))}
				</select>
			</div>
			<div>
				<select value={month.year()} onChange={e => onYearSelect(month, e.target.value)} className={styles.select}>
					{[...Array(this.props.numberOfYears)].map((value, index) => {
						const indexValue = moment().year() + index - (this.props.negative ? this.props.numberOfYears : 0);
						return (
							<option value={indexValue} key={indexValue}>
								{indexValue}
							</option>
						);
					})}
					{this.props.license &&
						[...Array(10)].map((value, index) => {
							const indexValue = moment().year() + index;
							return (
								<option value={indexValue} key={indexValue}>
									{indexValue}
								</option>
							);
						})}
				</select>
			</div>
		</div>
	)

	renderErrorBox = () => {
		const { error, valid } = this.state;
		const { errorMessage, isValid } = this.props;
		if (!valid || !isValid) {
			if (errorMessage && errorMessage.length > 0) {
				return <span styleName="error"> {errorMessage} </span>;
			}
			return <span styleName="error"> {error} </span>;
		}
		return null;
	}

	render() {
		const { placeholder, label, id, errorMessage } = this.props;
		const { date, error } = this.state;
		const navPrevIcon = <img src={directionArrow} alt="" styleName="icon prev" />;
		const navNextIcon = <img src={directionArrow} alt="" styleName="icon next" />;
		return (
			<div styleName={this.getClasses()}>
				<SingleDatePicker
					id={id}
					date={date}
					numberOfMonths={1}
					daySize={35}
					onDateChange={this.onDateChange}
					focused={this.state.focused}
					navPrev={navPrevIcon}
					navNext={navNextIcon}
					noBorder
					block
					isOutsideRange={this.props.isOutsideRange}
					hideKeyboardShortcutsPanel
					placeholder={placeholder}
					horizontalMonthPadding={5}
					onFocusChange={({ focused }) => this.setState({ focused })}
					renderMonthElement={this.renderMonthElement}
					isDayBlocked={this.props.isDayBlocked}
				/>
				{this.renderErrorBox()}
				{label && (
					<label styleName={classNames('input-label', { 'input-label--focused': this.state.focused || !date || error || errorMessage })} htmlFor={id}>
						{label}
					</label>
				)}
			</div>
		);
	}
}

export default CSSModules(CustomDatePicker, styles, { allowMultiple: true });
