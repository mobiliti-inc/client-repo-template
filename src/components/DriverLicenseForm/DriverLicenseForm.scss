@import "../../../styles/_variables.scss";

/* Add license info manually */
.container {
	@include signup-page-container;
	padding: 0 0 70px 0;
	display: flex;
	flex-direction: column;
	margin-top: 30px;
	min-height: 750px;

	@media (max-width: 640px) {
		min-height: 950px;
	}

	input {
		font-size: 20px;
	}

	.home {
		width: 100%;
	}

	.city,
	.state,
	.zip,
	.dob,
	.license,
	.expiration {
		margin-top: 30px;
	}

	.address {
		display: flex;
		flex-wrap: wrap;

		.city {
			width: 32%;

			@media (min-width: $tablet-width) {
				margin-right: 2%;
			}

			@media (max-width: $tablet-width) {
				width: 49%;
				margin-right: 1%;
			}

			@media (max-width: 640px) {
				width: 100%;
				margin-right: 0;
			}
		}

		.state {
			width: 32%;

			@media (min-width: $tablet-width) {
				margin-right: 2%;
			}

			@media (max-width: $tablet-width) {
				width: 49%;
				margin-left: 1%;
			}

			@media (max-width: 640px) {
				width: 100%;
				margin-left: 0;
			}
		}

		.zip {
			width: 32%;

			@media (max-width: $tablet-width) {
				width: 49%;
				margin-right: 1%;
			}

			@media (max-width: 640px) {
				width: 100%;
				margin-right: 0;
			}
		}

		.dob {
			width: 100%;

			@media (max-width: $tablet-width) {
				width: 49%;
				margin-left: 1%;
			}

			@media (max-width: 640px) {
				width: 100%;
				margin-left: 0;
			}
		}

		.license {
			width: 100%;
		}

		.expiration {
			width: 100%;
		}
	}

	.button {
		width: 100%;
		margin-top: 20px;

		@media (min-width: $mobile-width) {
			width: 160px;
		}
	}
}
