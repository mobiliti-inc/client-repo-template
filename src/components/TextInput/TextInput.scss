@import "../../styles/_variables.scss";

.input-group {
  position: relative;
  display: flex;
  flex-direction: column;
}

.input-base {
  height: 50px;
  width: 100%;
  font-size: 22px;
  font-family: inherit;
  border: none;
  transition: all .5s;
  border-bottom: 1px solid rgba(29, 49, 85, 0.2);
  text-transform: capitalize;
  font-family: $hanken-sans;

  &:focus {
    outline: none;
    border-bottom: 3px solid $blue;
  }

  &::-webkit-input-placeholder {
    color: $darker-blue;
    opacity: 0.6;
  }

  &:placeholder-shown+.input-label {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-40px);
  }

  &.error {
    border-bottom: 3px solid $fast-car-red;
  }
}

.input-email,
.input-password,
.input-number,
.input-image,
.input-text,
.input-file,
.input-date,
.input-url,
.input-tel {
  composes: input-base;
  text-transform: none;
}

.input-label {
  display: block;
  font-size: 12px;
  transform: translateY(-60px);
  color: $darker-blue;
  opacity: 0.6;
  font-family: $hanken-sans;
  line-height: 14px;
  text-transform: capitalize;
  transition: all .5s;
}

.checkmark-label {
  display: inline-block;
  position: absolute;
  top: 0px;
  right: 0px;
  height: 20px;
  width: 12px;
  border: solid $green-check;
  border-width: 0 2px 2px 0;
  cursor: pointer;
  transform: rotate(45deg);
}

.dark {
  color: $black;
  border-color: $black;

  &::placeholder {
    color: $black;
  }

  label {
    color: $black;
  }
}

.error-message {
  color: $fast-car-red;
  height: 14px;
  font-family: $hanken-sans;
  margin-top: 5px;
  font-size: 12px;
  line-height: 14px;
}

.password-visibility-indicator {
  display: inline-block;
  position: absolute;
  top: 15px;
  right: 0;
  line-height: 19px;
  color: $blue;
  font-family: $hanken-sans;
  font-size: 16px;
  border: none;
  background-color: transparent;
  text-transform: capitalize;
  cursor: pointer;

  &:active,
  &:focus {
    outline: none;
  }
}