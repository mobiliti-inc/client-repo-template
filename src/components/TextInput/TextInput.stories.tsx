import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import * as React from 'react';
import { TextInput } from '../';

storiesOf('TextInput', module)
	.add('default', () => (<div style={{ width: "300px" }}><TextInput type={text('type', 'text')}
		onChange={() => { }}
		id="text"
		value={text('value', '')}
		placeholder={text('value', '')}
		disabled={boolean('disabled', false)}
		isValid={boolean('isValid', false)}
		dark={boolean('dark', false)} /></div>
	)
	);