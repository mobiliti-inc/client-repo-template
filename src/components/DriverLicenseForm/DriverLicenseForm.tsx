
import React, { Component } from 'react';
import moment from 'moment';
// // Import React Script Library to load the Google object
import Script from 'react-load-script';

import TextInput, { INPUT_TYPES } from '../../components/TextInput/TextInput';
import Button from '../../components/Button/Button';
import CustomDatePicker from '../../components/CustomDatePicker/CustomDatePicker';

import { KEYS } from '../../../api';

import styles from './DriverLicenseForm.scss';

const defaultTextValue = {
	valid: true,
	error: '',
	value: ''
};

class DriverLicenseForm extends Component {
	static propTypes = {
		submitManualLicenseInfo: PropTypes.func.isRequired,
		showLoader: PropTypes.bool,
		subscriber: PropTypes.shape({
			fetched: PropTypes.bool,
			fetching: PropTypes.bool,
			address: PropTypes.shape({
				address1: PropTypes.string,
				address2: PropTypes.string,
				city: PropTypes.string,
				state: PropTypes.string,
				country: PropTypes.string,
				zip_code: PropTypes.string,
			}),
			driver_license_info: PropTypes.shape({
				driver_license_no: PropTypes.string,
				driver_license_expiry_date: PropTypes.string,
				driver_license_front_photo: PropTypes.string,
			}),
		}),
		buttonName: PropTypes.string,
		disabled: PropTypes.bool,
		validity: PropTypes.func,
		alertHandler: PropTypes.node,
	}

	static defaultProps = {
		showLoader: false,
		subscriber: null,
		buttonName: '',
		disabled: false,
		validity: () => {},
		alertHandler: null,
	}

	state = {
		fields: {
			address: defaultTextValue,
			city: defaultTextValue,
			state: defaultTextValue,
			zipCode: defaultTextValue,
			dateOfBirth: defaultTextValue,
			driverLicenseNumber: defaultTextValue,
			driverLicenseExpiryDate: defaultTextValue
		},
		inputChanged: false,
	}

	componentDidMount() {
		const { subscriber } = this.props;
		if (subscriber && subscriber.fetched) {
			this.populateFieldsWithInfo();
		}
	}

	componentDidUpdate = (prevProps, prevState) => {
		const { validity, } = this.props;
		const { inputChanged, fields } = this.state;
		if (inputChanged && prevState.inputChanged !== inputChanged) {
			validity(this.validity());
		}

		if (JSON.stringify(prevState.fields) !== JSON.stringify(fields) && inputChanged) {
			validity(this.validity());
		}
	}

	getUserData = () => {
		const {
			address,
			city,
			state,
			zipCode,
			driverLicenseNumber,
			driverLicenseExpiryDate,
			dateOfBirth
		} = this.state.fields;
		return {
			dateOfBirth: dateOfBirth.value,
			city: city.value,
			state: state.value,
			zipCode: zipCode.value,
			address: address.value,
			driverLicenseNumber: driverLicenseNumber.value,
			driverLicenseExpiryDate: driverLicenseExpiryDate.value,
		};
	}

	populateFieldsWithInfo = () => {
		const { subscriber } = this.props;
		this.setState(state => ({
			fields: {
				...state.fields,
				address: {
					valid: true,
					error: '',
					value: subscriber.address.address1 || '',
				},
				city: {
					valid: true,
					error: '',
					value: subscriber.address.city || '',
				},
				state: {
					valid: true,
					error: '',
					value: subscriber.address.state || '',
				},
				zipCode: {
					valid: true,
					error: '',
					value: subscriber.address.zip_code || '',
				},
				driverLicenseNumber: {
					valid: true,
					error: '',
					value: subscriber.driver_license_info.driver_license_no || '',
				},
				driverLicenseExpiryDate: {
					valid: true,
					error: '',
					value: subscriber.driver_license_info.driver_license_expiry_date || '',
				},
				dateOfBirth: {
					valid: true,
					error: '',
					value: subscriber.user.dob || '',
				}
			},
		}));
	}

	handleScriptLoad = () => {
		// Declare Options For Autocomplete
		const options = { types: ['geocode'] };

		// Initialize Google Autocomplete
		this.autocomplete = new window.google.maps.places.Autocomplete(
			document.getElementById('address-input'),
			options);
		// Fire event when a suggested name is selected
		this.autocomplete.addListener('place_changed',
			this.handlePlaceSelect);
	}

	handlePlaceSelect = () => {
		// Extract Values From Address Object
		const addressObject = this.autocomplete.getPlace();
		const address = addressObject.address_components;

		// Check if address is valid
		if (address) {
			// Set State
			this.setState(state => ({
				fields: {
					...state.fields,
					address: {
						valid: true,
						error: '',
						value: `${address[0] ? address[0].long_name : ''} ${address[1] ? address[1].long_name : ''}`
					},
					city: {
						valid: true,
						error: '',
						value: address[3] ? address[3].long_name : ''
					},
					state: {
						valid: true,
						error: '',
						value: address[5] ? address[5].long_name : ''
					},
					zipCode: {
						valid: true,
						error: '',
						value: address[7] ? address[7].long_name : ''
					},
				}
			}));
		}
	}

	checkDateFieldsValidity = (datesToValidate) => {
		/* eslint no-underscore-dangle: 0 */
		const isObjectValid = [];
		datesToValidate.map((date) => {
			const allFields = this.state.fields;
			if (date.value && moment(date.value, 'MM/DD/YYYY')._isValid) {
				allFields[date.stateName] = {
					value: date.value,
					error: '',
					valid: true,
				};
				isObjectValid.push(true);
			} else {
				allFields[date.stateName] = {
					value: date.value,
					error: 'Please enter a valid date',
					valid: false,
				};
				isObjectValid.push(false);
			}
			this.setState({ fields: allFields });
		});
		return isObjectValid.every((values) => values === true);
	}

	fieldValidator = (toValidate) => {
		const validatorRegex = /^[\w\d\s-.,]+$/;
		const { value } = toValidate;
		if (value === '') {
			return {
				value,
				valid: false,
				error: 'This field can not be empty'
			};
		} else if (!validatorRegex.test(value)) {
			return {
				value,
				valid: false,
				error: 'This field contains invalid characters'
			};
		} else if (value && value.length < 2) {
			return {
				value,
				valid: false,
				error: 'This field can not be less than 2 characters'
			};
		}
		return {
			value,
			error: '',
			valid: true
		};
	};

	validity = () => {
		const {
			address,
			city,
			state,
			zipCode,
			driverLicenseNumber,
			driverLicenseExpiryDate,
			dateOfBirth,
		} = this.state.fields;

		const validTextFields = this.checkInvalidFields([
			{ value: address.value, stateName: 'address' },
			{ value: city.value, stateName: 'city' },
			{ value: state.value, stateName: 'state' },
			{ value: zipCode.value, stateName: 'zipCode' },
			{ value: driverLicenseNumber.value, stateName: 'driverLicenseNumber' }
		]);
		const dateFieldsValid = this.checkDateFieldsValidity([
			{ value: dateOfBirth.value, stateName: 'dateOfBirth' },
			{ value: driverLicenseExpiryDate.value, stateName: 'driverLicenseExpiryDate' }
		]);

		return validTextFields && dateFieldsValid;
	}

	submitLisenseInfo = () => {
		if (this.validity()) this.props.submitManualLicenseInfo(this.getUserData());
	}

	checkInvalidFields = (toValidate) => {
		const isObjectValid = [];

		if (toValidate instanceof Array) {
			toValidate.map((valueToValidate) => {
				const handleObjectValidation = this.fieldValidator(valueToValidate);
				const allFields = this.state.fields;
				allFields[valueToValidate.stateName] = handleObjectValidation;
				this.setState({ fields: allFields });
				isObjectValid.push(handleObjectValidation.valid);
			});
			return isObjectValid.every((values) => values === true);
		}
		const { value, stateName } = toValidate;
		const handleValidation = this.fieldValidator(toValidate);
		this.setState({
			fields: {
				...this.state.fields,
				[stateName]: handleValidation
			}
		});
		return handleValidation.valid;
	}

	handleInputChange = ({ target: { name, value } }) => {
		this.setState(state => ({
			fields: {
				...state.fields,
				[name]: {
					valid: true,
					error: '',
					value
				}
			},
			inputChanged: true,
		}));
	}

	handleDatePicked = (date, state) => {
		if (date) {
			return this.setState({
				fields: {
					...this.state.fields,
					[state]: { ...date }
				},
				inputChanged: true,
			});
		}
		return null;
	}

	renderGooglePlacesScript = () => <Script url={`https://maps.googleapis.com/maps/api/js?key=${KEYS.GOOGLE_PLACES}&libraries=places`} onLoad={this.handleScriptLoad} />;

	render() {
		const {
			address,
			city,
			state,
			zipCode,
			dateOfBirth,
			driverLicenseNumber,
			driverLicenseExpiryDate,
		} = this.state.fields;

		const { showLoader, buttonName, disabled, alertHandler, } = this.props;
		return (
			<div styleName="container">
				{this.renderGooglePlacesScript()}
				<div styleName="home">
					<TextInput
						type={INPUT_TYPES.TEXT}
						onChange={this.handleInputChange}
						value={address.value}
						isValid={address.valid}
						name="address"
						label="Address"
						id="address-input"
						placeholder="Home Address"
						errorMessage={!address.valid ? address.error : null}
					/>
				</div>
				<div styleName="address">
					<div styleName="city">
						<TextInput
							type={INPUT_TYPES.TEXT}
							onChange={this.handleInputChange}
							value={city.value}
							isValid={city.valid}
							name="city"
							label="City"
							id="city-input"
							placeholder="City"
							errorMessage={!city.valid ? city.error : null}
						/>
					</div>
					<div styleName="state">
						<TextInput
							type={INPUT_TYPES.TEXT}
							onChange={this.handleInputChange}
							value={state.value}
							isValid={state.valid}
							name="state"
							label="State"
							id="state-input"
							placeholder="State"
							errorMessage={!state.valid ? state.error : null}
						/>
					</div>
					<div styleName="zip">
						<TextInput
							type={INPUT_TYPES.TEXT}
							onChange={this.handleInputChange}
							value={zipCode.value}
							isValid={zipCode.valid}
							name="zipCode"
							label="Zip"
							id="zip-input"
							placeholder="Zip"
							errorMessage={!zipCode.valid ? zipCode.error : null}
						/>
					</div>
					<div styleName="dob">
						<CustomDatePicker
							placeholder="Date of Birth"
							openToDate="1990/01/01"
							isValid={dateOfBirth.valid}
							errorMessage={!dateOfBirth.valid ? dateOfBirth.error : null}
							onDatePicked={(date) => this.handleDatePicked(date, 'dateOfBirth')}
							selectedDate={dateOfBirth.value}
							id="dob"
							label="Date of Birth"
							negative
						/>
					</div>
				</div>
				<div styleName="license">
					<TextInput
						type={INPUT_TYPES.TEXT}
						onChange={this.handleInputChange}
						value={driverLicenseNumber.value}
						isValid={driverLicenseNumber.valid}
						name="driverLicenseNumber"
						id="license-input"
						label="Driver's License Number"
						placeholder="Driver’s License Number"
						errorMessage={!driverLicenseNumber.valid ? driverLicenseNumber.error : null}
					/>
				</div>
				<div styleName="expiration">
					<CustomDatePicker
						placeholder="Expiration"
						openToDate="2020/01/01"
						isValid={driverLicenseExpiryDate.valid}
						errorMessage={!driverLicenseExpiryDate.valid ? driverLicenseExpiryDate.error : null}
						onDatePicked={(date) => this.handleDatePicked(date, 'driverLicenseExpiryDate')}
						selectedDate={driverLicenseExpiryDate.value}
						license
						id="expiration"
						label="Expiration Date"
					/>
				</div>
				{alertHandler}
				<div styleName="button">
					<Button
						showLoader={showLoader}
						onClick={this.submitLisenseInfo}
						customStyles={styles.button}
						ripple
						disabled={disabled}
					>
						{buttonName || 'Continue'}
					</Button>
				</div>
			</div>
		);
	}
}

export default CSSModules(DriverLicenseForm, styles, { allowMultiple: true });
