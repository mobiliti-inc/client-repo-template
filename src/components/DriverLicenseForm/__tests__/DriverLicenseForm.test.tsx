import Script from 'react-load-script';

import DriverLicenseForm from '../DriverLicenseForm';
import CustomDatePicker from '../../../components/CustomDatePicker/CustomDatePicker';
import TextInput from '../../../components/TextInput/TextInput';

import userModel from '../../../actions/sampleData/sampleUser';

describe('DriverLicenseForm', () => {
	const license = {
		frontImage: { image: '' },
		backImage: { image: '' },
	};

	const props = {
		submitManualLicenseInfo: jest.fn(),
		userData: {
			isLoading: false,
			auth: {
				accessToken: '',
				refreshToken: '',
			},
		},
		license,
		validity: jest.fn(),
		subscriber: {
			fetched: false,
			fetching: false,
			...userModel
		},
	};

	const fields = {
		address: {
			valid: true,
			error: [],
			value: 'Lorem'
		},
		city: {
			valid: true,
			error: [],
			value: 'Lorem'
		},
		state: {
			valid: true,
			error: [],
			value: 'Lorem'
		},
		zipCode: {
			valid: true,
			error: [],
			value: 'Lorem'
		},
		dateOfBirth: {
			valid: true,
			error: [],
			value: '01/01/1990',
		},
		driverLicenseNumber: {
			valid: true,
			error: [],
			value: '1234'
		},
		driverLicenseExpiryDate: {
			valid: true,
			error: [],
			value: '01/01/1990',
		}
	};

	jest.mock('moment', () => () => ({ format: () => '2018–01–30T12:34:56+00:00' }));

	const component = <DriverLicenseForm {...props} />;

	const tree = shallow(component);

	it('renders correctly', () => {
		matchSnapshot(tree);
	});

	it('should have two datepicker input', () => {
		expect(tree.find(CustomDatePicker)).toHaveLength(2);
	});

	it('should have all inputs', () => {
		expect(tree.find(CustomDatePicker)).toHaveLength(2);
		expect(tree.find(TextInput)).toHaveLength(5);
	});

	it('should render a google map script', () => {
		const mapScript = tree.find(Script);
		expect(mapScript).toHaveLength(1);
		expect(mapScript.props().url).toMatch('https://maps.googleapis.com/maps/api/js');
	});

	it('should have a function to check invalid fields', () => {
		const toValidate = { value: 'Lorem', stateName: 'textField' };
		const checkInvalidity = tree.instance().checkInvalidFields(toValidate);
		expect(checkInvalidity).toBeTruthy();
		expect(tree.state().fields.textField.valid).toBeTruthy();
	});

	it('should flag invalid fields, special chars and emptyness', () => {
		const subject = shallow(component);
		const specialCharInvalidField = { value: '@#$@%%@%$#', stateName: 'name' };
		const emptyField = { value: '', stateName: 'email' };
		const { checkInvalidFields } = subject.instance();
		expect(checkInvalidFields(specialCharInvalidField)).toBeFalsy();
		expect(subject.state('fields').name.valid).toBeFalsy();
		expect(subject.state('fields').name.error).toMatch('invalid characters');
		expect(checkInvalidFields(emptyField)).toBeFalsy();
		expect(subject.state('fields').email.valid).toBeFalsy();
		expect(subject.state('fields').email.error).toMatch('can not be empty');
	});

	it('should check invalid date fields', () => {
		const subject = shallow(component);
		const momentDate1 = '01/01/1990';
		const momentDate2 = '01/01/2020';
		const validDates = [
			{ value: momentDate1, stateName: 'date1' },
			{ value: momentDate2, stateName: 'date2' }
		];
		expect(subject.instance().checkDateFieldsValidity(validDates)).toBeTruthy();
	});

	it('should flag values less than 2 characters', () => {
		const twoCharsField = { stateName: 'twoChars', value: '1' };
		expect(tree.instance().checkInvalidFields(twoCharsField)).toBeFalsy();
		expect(tree.state().fields.twoChars.error).toMatch('This field can not be less than 2 characters');
	});

	it('should be able to validate objects of fields', () => {
		const objectToValidate = [
			{ stateName: 'lorem', value: 'Lorem' },
			{ stateName: 'ipsum', value: 'Ipsum' }
		];
		expect(tree.instance().checkInvalidFields(objectToValidate)).toBeTruthy();
	});

	it('should be able to submit license info', () => {
		const subject = shallow(component);
		subject.setState({ fields });
		subject.find('Button').simulate('click');
		expect(props.submitManualLicenseInfo).toHaveBeenCalledWith(subject.instance().getUserData());
	});

	it('should check validity on subscriber details missing', () => {
		const subject = shallow(<DriverLicenseForm {...props} />);
		expect(subject.state('inputChanged')).toBeFalsy();
		subject.setState({ inputChanged: true });
		expect(subject.instance().props.validity).toBeCalled();
	});

	it('should detect changes in field edit', () => {
		const subject = shallow(<DriverLicenseForm {...props} />);
		subject.setState({ fields });
		subject.setState({ fields: { ...fields, city: { value: 'Rome', valid: true, error: '' } } });
		expect(subject.instance().props.validity).toBeCalled();
	});

	it('should handle input changes', () => {
		const subject = shallow(<DriverLicenseForm {...props} />);
		subject.find('#city-input').simulate('change', { target: { value: 'Rome', name: 'city', placeholder: 'City' } });
		expect(subject.state('fields').city.valid).toBeTruthy();
		expect(subject.state('fields').city.value).toEqual('Rome');
	});
});
