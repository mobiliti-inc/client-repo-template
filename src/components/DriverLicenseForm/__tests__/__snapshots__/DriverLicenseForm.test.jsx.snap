// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DriverLicenseForm renders correctly 1`] = `
<div
  className="container"
>
  <Script
    attributes={Object {}}
    onCreate={[Function]}
    onError={[Function]}
    onLoad={[Function]}
    url="https://maps.googleapis.com/maps/api/js?key=IzaSyCudd566&libraries=places"
  />
  <div
    className="home"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage={null}
      id="address-input"
      inputDidChange={false}
      isValid={true}
      label="Address"
      name="address"
      onBlur={null}
      onChange={[Function]}
      placeholder="Home Address"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div
    className="address"
  >
    <div
      className="city"
    >
      <TextInput
        changePasswordVisibilityStatus={[Function]}
        className=""
        dark={false}
        disabled={false}
        errorMessage={null}
        id="city-input"
        inputDidChange={false}
        isValid={true}
        label="City"
        name="city"
        onBlur={null}
        onChange={[Function]}
        placeholder="City"
        shouldShowVisibility={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="state"
    >
      <TextInput
        changePasswordVisibilityStatus={[Function]}
        className=""
        dark={false}
        disabled={false}
        errorMessage={null}
        id="state-input"
        inputDidChange={false}
        isValid={true}
        label="State"
        name="state"
        onBlur={null}
        onChange={[Function]}
        placeholder="State"
        shouldShowVisibility={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="zip"
    >
      <TextInput
        changePasswordVisibilityStatus={[Function]}
        className=""
        dark={false}
        disabled={false}
        errorMessage={null}
        id="zip-input"
        inputDidChange={false}
        isValid={true}
        label="Zip"
        name="zipCode"
        onBlur={null}
        onChange={[Function]}
        placeholder="Zip"
        shouldShowVisibility={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="dob"
    >
      <CustomDatePicker
        errorMessage={null}
        id="dob"
        isDayBlocked={[Function]}
        isOutsideRange={[Function]}
        isValid={true}
        label="Date of Birth"
        license={false}
        negative={true}
        numberOfYears={99}
        onDatePicked={[Function]}
        openToDate="1990/01/01"
        placeholder="Date of Birth"
        selectedDate=""
      />
    </div>
  </div>
  <div
    className="license"
  >
    <TextInput
      changePasswordVisibilityStatus={[Function]}
      className=""
      dark={false}
      disabled={false}
      errorMessage={null}
      id="license-input"
      inputDidChange={false}
      isValid={true}
      label="Driver's License Number"
      name="driverLicenseNumber"
      onBlur={null}
      onChange={[Function]}
      placeholder="Driverâ€™s License Number"
      shouldShowVisibility={false}
      type="text"
      value=""
    />
  </div>
  <div
    className="expiration"
  >
    <CustomDatePicker
      errorMessage={null}
      id="expiration"
      isDayBlocked={[Function]}
      isOutsideRange={[Function]}
      isValid={true}
      label="Expiration Date"
      license={true}
      negative={false}
      numberOfYears={99}
      onDatePicked={[Function]}
      openToDate="2020/01/01"
      placeholder="Expiration"
      selectedDate=""
    />
  </div>
  <div
    className="button"
  >
    <Button
      bordered={false}
      buttonType="standard"
      className=""
      customStyles="button"
      disabled={false}
      icon={null}
      inverted={false}
      onClick={[Function]}
      ripple={true}
      showLoader={false}
    >
      Continue
    </Button>
  </div>
</div>
`;
